### ISSUE TITLE
Adversarial Generation of Continuous Images

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#137]
## Motivation

- Explore implicit neural representation (INR) based generative model.

## TL; DR

- Sample hyper-network parameters from the latent code.
- For efficiency, devise Factorized Multiplicative Modulation (FMM) and multi-scale INR.
- Interesting properties like grid extrapolation and out-of-the-box super resolution are observed.

## Previous Work

- INR-based generative model (GRAF, GIRAFFE, StyleNeRF)
- Hyper-network based approaches (GMDF)

## Method

![image](https://user-images.githubusercontent.com/46308407/145977081-4142d64d-1770-48a5-bccb-2fd2e3e07f65.png)

- Instead of directly sampling images through feed-forward operation, sample parameters of the hyper-network (coordinate MLP)

### Factorizable Multiplicative Modulation (FMM)

![image](https://user-images.githubusercontent.com/46308407/145977312-3af711ed-dfb1-41c9-80bb-622b2cbbf6c8.png)

- Directly producing all the parameters of the hyper-network is impractical (they use powerful hyper-network for better sample quality)
- Explore low rank factorization of the weight matrix
- Simple rank-1 decomposition leads to training instability, as many of the singular values become zero (~ spectral normalization)
- Instead, form a learnable weight matrix that is ***shared*** across samples (=template, equivalent of the learned constant vector in StyleGAN) and low-rank (rank=10) decompose the modulator matrix (element-wise multiplied after the application of sigmoid function)

### Multi-scale INR

![image](https://user-images.githubusercontent.com/46308407/145978333-e7b0f836-fbba-41e4-91b9-9c9bf4101ba0.png)

- Directly operating in high resolution (e.g., 1024x1024) is infeasible for hyper-network based INR-GAN.
- Hence, progressively increase the effective resolution just like conventional convolutional generators.
- For early blocks (=lower resolution), the coordinates are simply locally replicated (*nearest neighbor*) and queried to the MLP.
- This is equivalent to nearest neighbor upsampling during the feed forward. Only the last layer operates in full target resolution. 
- Starts at 64x64 and gradually rise to the target resolution.

## Experiments

![image](https://user-images.githubusercontent.com/46308407/145978381-ecf9218a-fa03-4642-ae5d-2d411aa55a1a.png)

- As they adopt StyleGAN2 configurations as baseline, mostly compare with StyleGAN2.

![image](https://user-images.githubusercontent.com/46308407/145978446-0d6b31f5-7c1c-40a5-865b-dbb8df9b5e1c.png)

- Higher rank gives advantages, but the gain diminishes. It shows that vanilla hypernetwork (=full rank) is clearly overfitting, or using excess capacity. (All parameters are different for all samples. Note that for unconditional synthesis task, there is large semantic common grounds for the samples, which is the intuition behind StyleGAN learnable constant tensor)

![image](https://user-images.githubusercontent.com/46308407/145978762-ac4ea2f1-a07d-4c4d-81ac-f9561546c36b.png)

- Some compromise with grid-based structure (such as bilinear interpolation) yields clear performance boosts, demonstrating the limitation of INR based models; it doesn't use spatial correlation or local context information due to pixel-independent structure.

![image](https://user-images.githubusercontent.com/46308407/145978956-f5de1348-009d-438f-9fb3-f6bf83d205bd.png)

- This in my personal opinion can potentially backfire as it shows the location bias of INR-based models. (Refer to StyleNeRF experiment on LIIF representation)

## Comments

- I was impressed by the paper *Generative Models as Distributions of Functions*. That work was more theoretical, while this one has made some compromises (e.g., convolutional discriminator) for better performance. Nevertheless, the performance shown by it is quite surprising, and there are some obvious upper hands with this model.
- One of the major advantage of this one is that they wisely modeled powerful hyper-network. With GMDF, the hyper-network was very shallow, probably due to computational burden, leading to very poor generation quality. (64x64 was their main resolution, if I remember correctly)
- The authors devised FMM (which I thought was very smart) and multi-scale INR framework to operate on higher dimensions, and the qualitative + quantitative evaluations give promising result. 
- For now, I do not see clear reason to employ INR in 2D modeling if not for 3D-aware synthesis purpose like GIRAFFE and StyleNeRF. But this line of works might lead to some pleasant surprises in the coming time.
